core/src/test/java/org/elasticsearch/cluster/ClusterChangedEventTests.java

```diff
@@ -0,0 +1,350 @@
+        // test when nodes have been removed between cluster states
+        ClusterState newState = createState(numNodesInCluster - 1, randomBoolean(), initialIndices);
+        ClusterChangedEvent event = new ClusterChangedEvent("_na_", newState, originalState);
+        assertTrue("Nodes should have been removed between cluster states", event.nodesRemoved());
+
+        // test when nodes have not been removed
+        newState = createState(numNodesInCluster, randomBoolean(), initialIndices);
+        event = new ClusterChangedEvent("_na_", newState, originalState);
+        assertFalse("Nodes should not have been removed between cluster states", event.nodesRemoved());
+
+        // test when nodes have only been added
+        newState = createState(numNodesInCluster + 1, randomBoolean(), initialIndices);
+        event = new ClusterChangedEvent("_na_", newState, originalState);
+        assertFalse("Nodes should not have been removed between cluster states", event.nodesRemoved());
+    }
+
+    /**
+     * Test the routing table changes checks.
+     */
+    public void testRoutingTableChanges() {
+        final int numNodesInCluster = 3;
+        final ClusterState originalState = createState(numNodesInCluster, randomBoolean(), initialIndices);
+
+        // routing tables and index routing tables are same object
+        ClusterState newState = ClusterState.builder(originalState).build();
+        ClusterChangedEvent event = new ClusterChangedEvent("_na_", originalState, newState);
+        assertFalse("routing tables should be the same object", event.routingTableChanged());
+        assertFalse("index routing table should be the same object", event.indexRoutingTableChanged(initialIndices.get(0)));
+
+        // routing tables and index routing tables aren't same object
+        newState = createState(numNodesInCluster, randomBoolean(), initialIndices);
+        event = new ClusterChangedEvent("_na_", originalState, newState);
+        assertTrue("routing tables should not be the same object", event.routingTableChanged());
+        assertTrue("index routing table should not be the same object", event.indexRoutingTableChanged(initialIndices.get(0)));
+
+        // index routing tables are different because they don't exist
+        newState = createState(numNodesInCluster, randomBoolean(), initialIndices.subList(1, initialIndices.size()));
+        event = new ClusterChangedEvent("_na_", originalState, newState);
+        assertTrue("routing tables should not be the same object", event.routingTableChanged());
+        assertTrue("index routing table should not be the same object", event.indexRoutingTableChanged(initialIndices.get(0)));
+    }
+
+    // Tests that the indices change list is correct as well as metadata equality when the metadata has changed.
+    private static void metaDataChangesCheck(final boolean changeClusterUUID) {
```

Comment:
@bleskes I'm not clear on what is meant here?  The `testMetaDataChangesOnNoMasterChange` and `testMetaDataChangesOnNewClusterUUID` tests each use this helper method as the test is essentially the same except for whether we change the cluster UUID or not.  I figured we needed a test to check metadata changes when the cluster UUID is changed and one when it is not.